### ISSUE TITLE
Experiences from trying to get GLMPCA to run with the book

### ISSUE THREAD

**Issue by LTLA** [kstreet13/scry#7]
Consider the following stretch of code:

```r
library(scRNAseq)
sce <- ZeiselBrainData()

library(scuttle)
sce <- logNormCounts(sce)

library(scran)
dec <- modelGeneVar(sce)
hvgs <- getTopHVGs(dec, n=1000)

library(scry)
out <- GLMPCA(sce[hvgs,], L=50, sz=sizeFactors(sce))
```

The most obvious issue is that I can't get it to work, `GLMPCA` conks out with:

```
Error in glmpca(Y = assay(object, assay), L = L, fam = fam, ctl = ctl,  : 
  Numerical divergence (deviance no longer finite), try increasing the penalty to improve stability of optimization.
```

Increasing the `penalty` to 100 causes it to run for a while. It's been 10 minutes at least; by comparison, running the usual approximate PCA on the equivalent log-expression matrix would take 1 second, maybe 2. I know it's doing more computational work than a vanilla PCA but this seems even slower than **zinbwave**. It's only a 3k x 1k matrix here, anything above 30 seconds would be too much.

There are also some other quality-of-life issues that you might consider addressing:

- Add a `subset=` argument to `GLMPCA()` so that we don't have to row-subset the entire SCE in the input, especially when only one of the matrices is of interest.
- Add a `SingleCellExperiment` method that defaults `sz=sizeFactors(x)`.
- What he said in #5.

**Comment by willtownes** [comment]
Hi Aaron, thanks for the feedback. The problems with numerical stability for large L and tuning issues of the penalty (+ computational speed) are all known issues with glmpca we are actively working on. I am pretty close to having something ready to release that scales better for larger L values. I am now testing it to see if it can work with large numbers of cells as well. I will post something here when I release it to CRAN, and any improvements should pass through to the scry interface automatically. Also I agree with all of your other suggestions. Sorry for the frustrations!

**Comment by willtownes** [comment]
by the way, we would recommend using the deviance for gene filtering rather than hvgs

**Comment by LTLA** [comment]
On what basis?

**Comment by willtownes** [comment]
it's more consistent with the multinomial model behind glm-pca. Also it gave better empirical results for eg clustering: https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1861-6/figures/5 , see also more comprehensive comparison by Germain et al where they recommend deviance for filtering: https://www.biorxiv.org/content/10.1101/2020.02.02.930578v3

However, this is totally optional and doesn't detract from your main points. You can use any filtering you choose as input to GLM-PCA. It's a separate issue from the speed and numerical problems.

**Comment by LTLA** [comment]
Seems like it would be tricky to choose an appropriate dispersion for the deviance. If you used a Poisson GLM, you'd just end up with a strong mean-dependent trend if the true dispersion was a constant non-zero value across all genes. Unless you're doing your own dispersion estimation, but then how would you do that before you've accounted for the structure in the GLM-PCA?

Yes, I also did read that paper. But they recommended imputation as well and I didn't know what to make of that, so I just left it alone and moved on.

**Comment by willtownes** [comment]
Hi Aaron, glmpca 0.2.0 has now been released. It should be faster and more numerically stable. Please try running your experiments again and let me know if things improve or if there are further problems. Thanks!

**Comment by LTLA** [comment]
Excellent. I will give it a spin sometime during the week.

**Comment by LTLA** [comment]
It finishes, which is good. The example in my original post took 262 seconds, which is... tolerable, for the purposes of putting it in the book. I haven't actually tested the output yet to see whether it solves the problem that I was hoping it would solve.

I'll test out the current state, which is definitely better than what it was before. Hopefully it turns out to do what I wanted; but if it does, it'll need to be a fair bit faster for mainstream use. Maybe a minute for 10k cells, which sounds pretty reasonable.

There's probably some low-hanging fruit with **BiocParallel**ization and **DelayedArray**ing, especially if you can make use of the parallelized `crossprod` provided by **DelayedArray** (via the **Matrix** generic). Or maybe there's some minibatching or some other approximation that I'm missing? I do notice a `glmpca:::create_minibatches` when I was poking around in there.

**Comment by willtownes** [comment]
Yes, minibatch is now supported. You can do `minibatch='memoized'` to use cached sufficient statistics to compute gradients or `minibatch='stochastic'` for stochastic gradient. The advantage of those methods is to save on memory, not necessarily to be faster, but anecdotally I have seen the stochastic gradient converge faster than full gradient (the default minibatch='none'). The problem with the stochastic gradient is it's hard to assess convergence due to the noise in the estimator, which causes it to often run for many excess iterations after it's already effectively converged. To set the minibatch size use `ctl=list(batch_size=1000)`. 

Also, for the Zeisel dataset it might be worth trying `fam='nb'` instead of the default `fam='poi'` because there are a couple of genes that have really huge counts compared to all the others, and negative binomial likelihood is a bit more numerically stable than Poisson in that scenario. 

Adding support for DelayedArray/ HDF5Array is on my to-do list ( willtownes/glmpca#18 ). Thanks for the suggestion about parallelizing the crossprod operation ( now included as willtownes/glmpca#22 )

**Comment by LTLA** [comment]
This is the call I've currently got in the book:

```r
sce.scry <- GLMPCA(sce.zeisel[top.hvgs,], fam="nb", 
    L=20, sz=sizeFactors(sce.zeisel))
```

One improvement would be to offer a subsetting argument so that I don't have to subset the SCE as input. This would ensure that the output SCE has the same dimensionality, with the only difference being a new redDim entry.

Another improvement would be to provide a SCE method where `sz=sizeFactors(object)` by default. 

The function is still achingly slow, though. I wonder whether there is more that can be done to speed it up. For example, using the PCA on the log-values to provide the initial `U` and `V` when `init=` is not supplied, maybe that would get start you closer to the target space. Or sharing information across iterations with different learning rates. 

(I don't really understand why multiple learning rates are even required; I'm assuming the learning rate is somehow related to the step size per iteration, but surely this could be dynamically adjusted within a run rather than restarting the run with a different rate? We use a similar approach in **edgeR** to guarantee that each step decreases deviance.)

**Comment by willtownes** [comment]
thanks for these good suggestions.
